crates/types/src/address.rs

```diff
@@ -138,15 +150,40 @@ impl From<Addr> for String {
     }
 }
 
+// Convert the raw bytes bo checksumed hex string according to ERC-55:
+// https://eips.ethereum.org/EIPS/eip-55#implementation
+//
+// Adapted from alloy-rs:
+// https://github.com/alloy-rs/core/blob/v0.7.7/crates/primitives/src/bits/address.rs#L294-L320
 impl fmt::Display for Addr {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{}{}", Self::PREFIX, self.0)
+        let mut checksum = [0; 40];
+        // This encodes hex in lowercase.
+        hex::encode_to_slice(&self.0, &mut checksum).unwrap();
+
+        let mut hasher = Keccak256::new();
+        // Note we're hashing the UTF-8 hex string, not the raw bytes.
+        hasher.update(checksum);
+        let hash = hasher.finalize();
```

Comment:
Here we do the keccak hash in the contract instead of using `ctx.api.keccak_256`. However according to our benchmark, there isn't much performance difference. @Rhaki